Bearded.Utilities/Monads/Result.cs

```diff
@@ -0,0 +1,128 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Bearded.Utilities.Monads
+{
+    public readonly struct Result<TResult, TError> : IEquatable<Result<TResult, TError>>
+    {
+        private readonly bool isSuccess;
+        private readonly TResult result;
+        private readonly TError error;
+
+        private Result(bool isSuccess, TResult result, TError error)
+        {
+            this.isSuccess = isSuccess;
+            this.result = result;
+            this.error = error;
+        }
+
+        public TResult ResultOrDefault(TResult @default) => isSuccess ? result : @default;
+
+        public TResult ResultOrDefault(Func<TResult> defaultProvider) => isSuccess ? result : defaultProvider();
+
+        public TResult ResultOrThrow(Func<TError, Exception> exceptionProvider) =>
+            isSuccess ? result : throw exceptionProvider(error);
+
+        public Result<TOut, TError> Select<TOut>(Func<TResult, TOut> selector) =>
+            isSuccess ? (Result<TOut, TError>) Result.Success(selector(result)) : Result.Failure(error);
+
+        public Result<TOut, TError> SelectMany<TOut>(Func<TResult, Result<TOut, TError>> selector) =>
+            isSuccess ? selector(result) : Result.Failure(error);
+
+        public void Match(Action<TResult> onSuccess)
+        {
+            if (isSuccess)
+            {
+                onSuccess(result);
+            }
+        }
+
+        public void Match(Action<TResult> onSuccess, Action<TError> onFailure)
```

Comment:
Something that worries my slightly is that default(Result<,>) will have error == null, which could lead to a NullReferenceException in onFailure here and below.

Reply:
That's a good point.

Do you have any suggestions on how to resolve this? I can't think of a good way to avoid.

Reply:
The only thing we can do is catching default state ourselves and throwing (can't check for null, since someone might want to Result<T, MyError?>, would need to have another flag, and check that all over the place). I'm not sure if that's how we want this to behave? Ideally, structs have meaningful default values... in this case that would only work if we enforce TError to be nullable, which isn't great.

Reply:
We could check it everywhere, but what would we do if we found that the result is neither success nor failure? Throw an exception? That doesn't really get us much further. Do we add a third "onDefault" to all match functions? That also seems weird.

My preference would be to leave as-is for the time being, because any solution I can think of complicates the type, without really providing a satisfying solution.

Reply:
Agreed. Though is there a way for us to mark error as nullable or something? I guess not, given that it can be a valuetype? Perhaps there is an attribute we can use?

Reply:
Please take a look at the proposed changes based on some added annotations.